File: SDK/NodeGraph/rust/tests/runtime_tests.rs
use std::collections::HashSet;
use std::sync::{Arc, Mutex};
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use async_trait::async_trait;
use httpmock::Method::POST;
use httpmock::MockServer;
use nodegraph_sdk::{
    CreateSessionRequest, NodeDefinition, NodeGraphClient, NodeGraphDocument, NodeGraphEdge,
    NodeGraphExecutionOptions, NodeGraphNode, NodeGraphRuntime, NodeGraphRuntimeClient,
    NodeGraphRuntimeOptions, NodeGraphViewport, NodeLibraryEnvelope, NodeLibraryFieldDefinition,
    NodePortDefinition, Position, RuntimeRegistrationRequest, RuntimeRegistrationResponse,
};
use serde_json::{Map, Value, json};

fn json_map(value: Value) -> Map<String, Value> {
    value.as_object().cloned().expect("expected json object")
}

fn create_hello_runtime(
    now: Option<Arc<dyn Fn() -> SystemTime + Send + Sync>>,
) -> NodeGraphRuntime {
    let mut runtime = NodeGraphRuntime::new(NodeGraphRuntimeOptions {
        domain: "hello-world".into(),
        client_name: Some("Hello Runtime Host".into()),
        control_base_url: "http://127.0.0.1:4310/runtime".into(),
        library_version: "hello-world@1".into(),
        capabilities: None,
        runtime_id: Some("rt_demo_001".into()),
        cache_ttl: Some(Duration::from_secs(30 * 60)),
        now,
    })
    .expect("runtime should be created");

    runtime.register_node(
        NodeDefinition::new(
            "greeting_source",
            "Greeting Source",
            "Hello World",
            |context| {
                let name = context
                    .values
                    .get("name")
                    .and_then(Value::as_str)
                    .unwrap_or("World");
                context.emit("text", json!(format!("Hello, {name}!")));
                Ok(())
            },
        )
        .with_description("Create the base greeting text.")
        .with_outputs(vec![NodePortDefinition {
            id: "text".into(),
            label: "Text".into(),
            data_type: Some("hello/text".into()),
        }])
        .with_fields(vec![NodeLibraryFieldDefinition {
            key: "name".into(),
            label: "Name".into(),
            placeholder: None,
            kind: "text".into(),
            default_value: Some(json!("World")),
            options_endpoint: None,
        }]),
    );

    runtime.register_node(
        NodeDefinition::new(
            "console_output",
            "Console Output",
            "Hello World",
            |context| {
                if let Some(value) = context.read_input("text") {
                    context.push_result("console", value);
                }
                Ok(())
            },
        )
        .with_description("Write the greeting to the host result buffer.")
        .with_inputs(vec![NodePortDefinition {
            id: "text".into(),
            label: "Text".into(),
            data_type: Some("hello/text".into()),
        }]),
    );

    runtime
}

fn create_hello_graph() -> NodeGraphDocument {
    NodeGraphDocument {
        graph_id: None,
        name: "Hello World".into(),
        description: None,
        nodes: vec![
            NodeGraphNode {
                id: "node_source".into(),
                node_type: "default".into(),
                position: Position { x: 0.0, y: 0.0 },
                data: json_map(json!({
                    "label": "Greeting Source",
                    "nodeType": "greeting_source",
                    "outputs": [
                        {
                            "id": "text",
                            "label": "Text",
                            "dataType": "hello/text"
                        }
                    ],
                    "values": {
                        "name": "Codex"
                    }
                })),
                width: None,
                height: None,
                style: None,
            },
            NodeGraphNode {
                id: "node_output".into(),
                node_type: "default".into(),
                position: Position { x: 280.0, y: 0.0 },
                data: json_map(json!({
                    "label": "Console Output",
                    "nodeType": "console_output",
                    "inputs": [
                        {
                            "id": "text",
                            "label": "Text",
                            "dataType": "hello/text"
                        }
                    ],
                    "values": {}
                })),
                width: None,
                height: None,
                style: None,
            },
        ],
        edges: vec![NodeGraphEdge {
            id: "edge_source_output".into(),
            source: "node_source".into(),
            target: "node_output".into(),
            source_handle: Some("text".into()),
            target_handle: Some("text".into()),
            label: None,
            edge_type: None,
            animated: None,
            invalid_reason: None,
        }],
        viewport: NodeGraphViewport {
            x: 0.0,
            y: 0.0,
            zoom: 1.0,
        },
    }
}

struct RecordingRuntimeClient {
    calls: Arc<Mutex<Vec<RuntimeRegistrationRequest>>>,
}

#[async_trait]
impl NodeGraphRuntimeClient for RecordingRuntimeClient {
    async fn register_runtime(
        &self,
        request: &RuntimeRegistrationRequest,
    ) -> Result<RuntimeRegistrationResponse, nodegraph_sdk::NodeGraphError> {
        self.calls.lock().expect("lock").push(request.clone());
        Ok(RuntimeRegistrationResponse {
            runtime_id: request.runtime_id.clone(),
            cached: false,
            expires_at: "2026-03-21T00:30:00.000Z".into(),
            library_version: request.library_version.clone(),
        })
    }
}

#[tokio::test]
async fn client_registers_runtime_and_creates_session_with_runtime_id() {
    let server = MockServer::start_async().await;
    let register_mock = server
        .mock_async(|when, then| {
            when.method(POST).path("/api/sdk/runtimes/register");
            then.status(201).json_body(json!({
                "runtimeId": "rt_demo_001",
                "cached": false,
                "expiresAt": "2026-03-21T00:30:00.000Z",
                "libraryVersion": "hello-world@1"
            }));
        })
        .await;
    let session_mock = server
        .mock_async(|when, then| {
            when.method(POST)
                .path("/api/sdk/sessions")
                .json_body(json!({
                    "runtimeId": "rt_demo_001",
                    "completionWebhook": "http://127.0.0.1:4310/api/completed"
                }));
            then.status(201).json_body(json!({
                "sessionId": "ngs_demo",
                "runtimeId": "rt_demo_001",
                "editorUrl": format!("{}/editor/ngs_demo", server.base_url()),
                "accessType": "private"
            }));
        })
        .await;
    let client = NodeGraphClient::new(server.base_url());

    client
        .register_runtime(&RuntimeRegistrationRequest {
            runtime_id: "rt_demo_001".into(),
            domain: "hello-world".into(),
            client_name: None,
            control_base_url: "http://127.0.0.1:4310/runtime".into(),
            library_version: "hello-world@1".into(),
            capabilities: None,
            library: NodeLibraryEnvelope::default(),
        })
        .await
        .expect("register runtime should succeed");

    client
        .create_session(&CreateSessionRequest {
            runtime_id: "rt_demo_001".into(),
            completion_webhook: "http://127.0.0.1:4310/api/completed".into(),
            graph: None,
            metadata: None,
        })
        .await
        .expect("create session should succeed");

    register_mock.assert_async().await;
    session_mock.assert_async().await;
}

#[tokio::test]
async fn runtime_skips_redundant_registration_within_cache_ttl() {
    let current_time = Arc::new(Mutex::new(UNIX_EPOCH + Duration::from_secs(1_742_516_800)));
    let now = {
        let current_time = Arc::clone(&current_time);
        Arc::new(move || *current_time.lock().expect("lock"))
    };
    let mut runtime = create_hello_runtime(Some(now));
    let calls = Arc::new(Mutex::new(Vec::<RuntimeRegistrationRequest>::new()));
    let client = RecordingRuntimeClient {
        calls: Arc::clone(&calls),
    };

    runtime
        .ensure_registered(&client, false)
        .await
        .expect("first registration should succeed");
    runtime
        .ensure_registered(&client, false)
        .await
        .expect("cached registration should succeed");
    *current_time.lock().expect("lock") += Duration::from_secs(31 * 60);
    runtime
        .ensure_registered(&client, false)
        .await
        .expect("expired registration should succeed");

    let calls = calls.lock().expect("lock");
    assert_eq!(calls.len(), 2);
    assert_eq!(calls[0].runtime_id, runtime.runtime_id());
    assert_eq!(calls[1].runtime_id, runtime.runtime_id());
    assert_eq!(
        runtime
            .get_library()
            .nodes
            .iter()
            .map(|node| node.node_type.clone())
            .collect::<Vec<_>>(),
        vec!["greeting_source".to_string(), "console_output".to_string()]
    );
}

#[test]
fn runtime_executes_hello_graph_and_captures_profiling() {
    let runtime = create_hello_runtime(None);

    let result = runtime.execute_graph(&create_hello_graph(), None);

    assert_eq!(result.status, "completed");
    assert_eq!(result.results["console"], vec![json!("Hello, Codex!")]);
    assert_eq!(result.profiler["node_source"].call_count, 1);
    assert_eq!(result.profiler["node_output"].call_count, 1);
    assert!(result.profiler["node_output"].total_duration_ms >= 0.0);
}

#[test]
fn runtime_debugger_pauses_on_breakpoint_and_can_continue() {
    let runtime = create_hello_runtime(None);
    let mut debug_session = runtime.create_debugger(
        &create_hello_graph(),
        Some(NodeGraphExecutionOptions {
            breakpoints: Some(HashSet::from(["node_output".to_string()])),
            max_steps: None,
            max_wall_time: None,
        }),
    );

    let first_step = debug_session.step();
    assert_eq!(first_step.status, "paused");
    assert_eq!(
        first_step
            .last_event
            .expect("last event should exist")
            .node_id,
        "node_source"
    );

    let paused = debug_session.continue_execution();
    assert_eq!(paused.status, "paused");
    assert_eq!(paused.pause_reason.as_deref(), Some("breakpoint"));
    assert_eq!(paused.pending_node_id.as_deref(), Some("node_output"));

    let completed = debug_session.continue_execution();
    assert_eq!(completed.status, "completed");
    assert_eq!(completed.results["console"], vec![json!("Hello, Codex!")]);
    assert_eq!(completed.profiler["node_output"].call_count, 1);
}

File: SDK/NodeGraph/dotnet/NodeGraphSdk.Tests/RuntimeTests.cs
using System.Net;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using NodeGraphSdk;
using Xunit;

namespace NodeGraphSdk.Tests;

public sealed class RuntimeTests
{
    private sealed class StubHttpMessageHandler(Func<HttpRequestMessage, HttpResponseMessage> handler) : HttpMessageHandler
    {
        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            return Task.FromResult(handler(request));
        }
    }

    private static NodeGraphRuntime CreateHelloRuntime(Func<DateTimeOffset>? nowProvider = null)
    {
        var runtime = new NodeGraphRuntime(new NodeGraphRuntimeOptions
        {
            Domain = "hello-world",
            ClientName = "Hello Runtime Host",
            ControlBaseUrl = "http://127.0.0.1:4310/runtime",
            LibraryVersion = "hello-world@1",
            NowProvider = nowProvider,
        });

        runtime.RegisterNode(new NodeDefinition
        {
            Type = "greeting_source",
            DisplayName = "Greeting Source",
            Description = "Create the base greeting text.",
            Category = "Hello World",
            Outputs =
            [
                new NodePortDefinition
                {
                    Id = "text",
                    Label = "Text",
                    DataType = "hello/text",
                },
            ],
            Fields =
            [
                new NodeLibraryFieldDefinition
                {
                    Key = "name",
                    Label = "Name",
                    Kind = "text",
                    DefaultValue = "World",
                },
            ],
            ExecuteAsync = context =>
            {
                context.Emit("text", $"Hello, {context.Values["name"]}!");
                return Task.CompletedTask;
            },
        });

        runtime.RegisterNode(new NodeDefinition
        {
            Type = "console_output",
            DisplayName = "Console Output",
            Description = "Write the greeting to the host result buffer.",
            Category = "Hello World",
            Inputs =
            [
                new NodePortDefinition
                {
                    Id = "text",
                    Label = "Text",
                    DataType = "hello/text",
                },
            ],
            ExecuteAsync = context =>
            {
                context.PushResult("console", context.ReadInput("text"));
                return Task.CompletedTask;
            },
        });

        return runtime;
    }

    private static NodeGraphDocument CreateHelloGraph()
    {
        return new NodeGraphDocument
        {
            Name = "Hello World",
            Nodes =
            [
                new NodeGraphNode
                {
                    Id = "node_source",
                    Type = "default",
                    Position = new Position(),
                    Data = new Dictionary<string, object?>
                    {
                        ["label"] = "Greeting Source",
                        ["nodeType"] = "greeting_source",
                        ["outputs"] = new[]
                        {
                            new Dictionary<string, object?>
                            {
                                ["id"] = "text",
                                ["label"] = "Text",
                                ["dataType"] = "hello/text",
                            },
                        },
                        ["values"] = new Dictionary<string, object?>
                        {
                            ["name"] = "Codex",
                        },
                    },
                },
                new NodeGraphNode
                {
                    Id = "node_output",
                    Type = "default",
                    Position = new Position(),
                    Data = new Dictionary<string, object?>
                    {
                        ["label"] = "Console Output",
                        ["nodeType"] = "console_output",
                        ["inputs"] = new[]
                        {
                            new Dictionary<string, object?>
                            {
                                ["id"] = "text",
                                ["label"] = "Text",
                                ["dataType"] = "hello/text",
                            },
                        },
                        ["values"] = new Dictionary<string, object?>(),
                    },
                },
            ],
            Edges =
            [
                new NodeGraphEdge
                {
                    Id = "edge_source_output",
                    Source = "node_source",
                    SourceHandle = "text",
                    Target = "node_output",
                    TargetHandle = "text",
                },
            ],
            Viewport = new NodeGraphViewport(),
        };
    }

    [Fact]
    public async Task ClientRegistersRuntimeAndCreatesSessionWithRuntimeIdAsync()
    {
        var requests = new List<HttpRequestMessage>();
        using var httpClient = new HttpClient(new StubHttpMessageHandler(request =>
        {
            requests.Add(request);

            if (request.RequestUri!.AbsolutePath.EndsWith("/api/sdk/runtimes/register", StringComparison.Ordinal))
            {
                return new HttpResponseMessage(HttpStatusCode.Created)
                {
                    Content = new StringContent(
                        "{\"runtimeId\":\"rt_demo_001\",\"cached\":false,\"expiresAt\":\"2026-03-21T00:30:00.000Z\",\"libraryVersion\":\"hello-world@1\"}",
                        Encoding.UTF8,
                        "application/json"),
                };
            }

            return new HttpResponseMessage(HttpStatusCode.Created)
            {
                Content = new StringContent(
                    "{\"sessionId\":\"ngs_demo\",\"runtimeId\":\"rt_demo_001\",\"editorUrl\":\"http://localhost:3000/editor/ngs_demo\",\"accessType\":\"private\"}",
                    Encoding.UTF8,
                    "application/json"),
            };
        }));
        var client = new NodeGraphClient(httpClient, "http://localhost:3000");

        await client.RegisterRuntimeAsync(new RuntimeRegistrationRequest
        {
            RuntimeId = "rt_demo_001",
            Domain = "hello-world",
            ControlBaseUrl = "http://127.0.0.1:4310/runtime",
            LibraryVersion = "hello-world@1",
            Library = new NodeLibraryEnvelope(),
        });

        await client.CreateSessionAsync(new CreateSessionRequest
        {
            RuntimeId = "rt_demo_001",
            CompletionWebhook = "http://127.0.0.1:4310/api/completed",
        });

        Assert.Equal(2, requests.Count);
        Assert.EndsWith("/api/sdk/runtimes/register", requests[0].RequestUri!.AbsolutePath, StringComparison.Ordinal);
        Assert.EndsWith("/api/sdk/sessions", requests[1].RequestUri!.AbsolutePath, StringComparison.Ordinal);
    }

    [Fact]
    public async Task RuntimeSkipsRedundantRegistrationWithinCacheTtlAsync()
    {
        var currentTime = new DateTimeOffset(2026, 3, 21, 0, 0, 0, TimeSpan.Zero);
        var runtime = CreateHelloRuntime(() => currentTime);
        var calls = new List<RuntimeRegistrationRequest>();
        var client = new RecordingRuntimeClient(request =>
        {
            calls.Add(request);
            return new RuntimeRegistrationResponse
            {
                RuntimeId = request.RuntimeId,
                Cached = calls.Count > 1,
                ExpiresAt = "2026-03-21T00:30:00.000Z",
                LibraryVersion = request.LibraryVersion,
            };
        });

        await runtime.EnsureRegisteredAsync(client);
        await runtime.EnsureRegisteredAsync(client);
        currentTime = currentTime.AddMinutes(31);
        await runtime.EnsureRegisteredAsync(client);

        Assert.Equal(2, calls.Count);
        Assert.Equal(runtime.RuntimeId, calls[0].RuntimeId);
        Assert.Equal(runtime.RuntimeId, calls[1].RuntimeId);
        Assert.Equal(["greeting_source", "console_output"], runtime.GetLibrary().Nodes.Select(node => node.Type).ToArray());
    }

    [Fact]
    public async Task RuntimeExecutesHelloGraphAndCapturesProfilingAsync()
    {
        var runtime = CreateHelloRuntime();

        var result = await runtime.ExecuteGraphAsync(CreateHelloGraph());

        Assert.Equal("completed", result.Status);
        Assert.Equal(["Hello, Codex!"], result.Results["console"].Cast<string>().ToArray());
        Assert.Equal(1, result.Profiler["node_source"].CallCount);
        Assert.Equal(1, result.Profiler["node_output"].CallCount);
        Assert.True(result.Profiler["node_output"].TotalDurationMs >= 0);
    }

    [Fact]
    public async Task RuntimeDebuggerPausesOnBreakpointAndCanContinueAsync()
    {
        var runtime = CreateHelloRuntime();
        var debugSession = runtime.CreateDebugger(CreateHelloGraph(), new NodeGraphExecutionOptions
        {
            Breakpoints = ["node_output"],
        });

        var firstStep = await debugSession.StepAsync();
        Assert.Equal("paused", firstStep.Status);
        Assert.Equal("node_source", firstStep.LastEvent?.NodeId);

        var paused = await debugSession.ContinueAsync();
        Assert.Equal("paused", paused.Status);
        Assert.Equal("breakpoint", paused.PauseReason);
        Assert.Equal("node_output", paused.PendingNodeId);

        var completed = await debugSession.ContinueAsync();
        Assert.Equal("completed", completed.Status);
        Assert.Equal(["Hello, Codex!"], completed.Results["console"].Cast<string>().ToArray());
        Assert.Equal(1, completed.Profiler["node_output"].CallCount);
    }

    [Fact]
    public async Task RuntimeDebuggerCanReplaceBreakpointsWhileReusingTheSameSessionAsync()
    {
        var runtime = CreateHelloRuntime();
        var debugSession = runtime.CreateDebugger(CreateHelloGraph());

        var firstStep = await debugSession.StepAsync();
        Assert.Equal("paused", firstStep.Status);
        Assert.Equal("node_output", firstStep.PendingNodeId);

        debugSession.SetBreakpoints(["node_output"]);

        var paused = await debugSession.ContinueAsync();
        Assert.Equal("paused", paused.Status);
        Assert.Equal("breakpoint", paused.PauseReason);
        Assert.Equal("node_output", paused.PendingNodeId);

        debugSession.SetBreakpoints([]);

        var completed = await debugSession.ContinueAsync();
        Assert.Equal("completed", completed.Status);
        Assert.Equal(["Hello, Codex!"], completed.Results["console"].Cast<string>().ToArray());
    }

    private sealed class RecordingRuntimeClient(Func<RuntimeRegistrationRequest, RuntimeRegistrationResponse> handler) : INodeGraphRuntimeClient
    {
        public Task<RuntimeRegistrationResponse> RegisterRuntimeAsync(RuntimeRegistrationRequest request, CancellationToken cancellationToken = default)
        {
            return Task.FromResult(handler(request));
        }
    }
}
